fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,411 @@
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, FitRes]],
+        failures: List[Union[Tuple[ClientProxy, FitRes], BaseException]],
+    ) -> Tuple[Optional[Parameters], Dict[str, Scalar]]:
+        """
+        Aggregate fit results by weighing them against the adjustment weights and then summing them.
+
+        Collects the fit metrics that will be used to change the adjustment weights for the next round.
+
+        Args:
+            server_round: (int) the current server round.
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' fit results.
+            failures: (List[Union[Tuple[ClientProxy, FitRes], BaseException]]) the clients' fit failures.
+
+        Returns:
+            (Tuple[Optional[Parameters], Dict[str, Scalar]]) A tuple containing the aggregated parameters
+                and the aggregated fit metrics.
+        """
+        # The original aggregated parameters is done by the super class (which we want to
+        # override its behaviour here), so we are discarding it to recalculate them in the lines below
+        _, metrics_aggregated = super().aggregate_fit(server_round, results, failures)
+
+        self.train_metrics = {}
+        for client_proxy, fit_res in results:
+            self.train_metrics[client_proxy.cid] = fit_res.metrics
+
+        parameters_aggregated = ndarrays_to_parameters(self.weight_and_aggregate_results(results))
+
+        return parameters_aggregated, metrics_aggregated
+
+    def aggregate_evaluate(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, EvaluateRes]],
+        failures: List[Union[Tuple[ClientProxy, EvaluateRes], BaseException]],
+    ) -> Tuple[Optional[float], Dict[str, Scalar]]:
+        """
+        Aggregate evaluation losses using weighted average.
+
+        Collects the evaluation metrics and updates the adjustment weights, which will be used
+        when aggregating the results for the next round.
+
+        Args:
+            server_round: (int) the current server round.
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' evaluate results.
+            failures: (List[Union[Tuple[ClientProxy, FitRes], BaseException]]) the clients' evaluate failures.
+
+        Returns:
+            (Tuple[Optional[float], Dict[str, Scalar]]) A tuple containing the aggregated evaluation loss
+                and the aggregated evaluation metrics.
+        """
+        loss_aggregated, metrics_aggregated = super().aggregate_evaluate(server_round, results, failures)
+
+        self.evaluation_metrics = {}
+        for client_proxy, eval_res in results:
+            cid = client_proxy.cid
+            self.evaluation_metrics[cid] = eval_res.metrics
+            # adding the loss to the metrics
+            val_loss_key = FairnessMetricType.LOSS.value
+            assert loss_aggregated is not None
+            self.evaluation_metrics[cid][val_loss_key] = loss_aggregated
```

Comment:
I'm a little bit confused as to why we're adding in the aggregated loss here rather than the individual loss values on each client after weight aggregation? That is, I believe we should be storing $G_{\hat{D}_i}(\theta^r)$ here for each individual $i$, rather than their weighted average of these values.

Reply:
Perhaps I'm missing something here though.

Reply:
The refactor would be as simple as using `eval_res.loss` here instead of `loss_aggregated` I think

Reply:
That's right, thanks for catching it.